std/stdio.d

```diff
@@ -651,15 +652,15 @@ Detaches from the underlying file. If the sole owner, calls $(D close).
 
 Throws: $(D ErrnoException) on failure if closing the file.
   */
-    void detach() @safe
+    void detach() @trusted
     {
         if (!_p) return;
-        if (_p.refs == 1)
+        if (atomicLoad(_p.refs) == 1)
             close();
         else
```

Comment:
I think a rearrangement of what is done is in order:

1. You need a private function that closes the file without detaching.
2. close needs to call private func, and then detach
3. All RC should be done in detach.